### PR TITLE
Add dev and prod Dockerfiles

### PR DESCRIPTION
**Pull request by macropin**
Added Dockerfiles suitable for development and official release building.

And a Makefile with targets for automation of docker image building.

Both may require some fine tuning, but should be a good starting point.

Refs: Luminarys/synapse/issues/74

### CODE REVIEW

**Comment by Luminarys** [comment]
Major issues:
* ports need to be opened to allow for UDP/incoming communication to the container
* config is completely default
* no file mount for downloading

I'm a bit unsure of the value overall of providing these images, as they have to be a bit too opinionated (if you let the user specify the config for instance, the ports may change), and given the low dependency count of synapse probably don't provide much more value than compiling it yourself.

**Comment by aaomidi** [comment]
How would the port being different affect the user? It's common practice to allow mounting of a "config" folder where the user has access to the configurations of the program from the host machine.

The port is handled independently through Docker, so the user would be able to specify which port in config and subsequently forward that port through Docker.

**Comment by ddevault** [comment]
I'm of the opinion that a Dockerfile is not something we want to make "official", but I encourage you to make one separately. The official way to install synapse should be only through your distribution's package manager.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  && useradd --comment "Synapse" --home ${SYNAPSE_HOME} --shell /usr/sbin/nologin --system synapse \
+  && mkdir -p ${SYNAPSE_HOME} \
+  && chown synapse:synapse ${SYNAPSE_HOME} \
+  ;
+
+COPY --from=build /usr/src/target/release/synapse /usr/src/target/release/sycli /usr/local/bin/
+
+USER synapse
+WORKDIR $SYNAPSE_HOME
+CMD ["synapse"]
```
Comment (ddevault) [inline Dockerfile:30]: This should probably `FROM scratch` and copy over only the files necessary to run Synapse

Comment (macropin) [reply Dockerfile:30]: That'll only work if synapse can be statically compiled. Is that possible?

Comment (ddevault) [reply Dockerfile:30]: It'll work even if you don't statically link synapse. You just copy in its dependencies, too. But yes, you can statically link synapse.

**Comment by macropin** [comment]
As I said in above these are a starting point (and a usable one at that). A decision should still be made on the approach for configuring the container:

> ports need to be opened to allow for UDP/incoming communication to the container

Default ports can be `EXPOSED` in the Dockerfile, alternatively, users can expose the ports they require with `-p`.

> config is completely default

There are three main approaches to handling config:
  1. User mounts a static config file into the container in the appropriate place using `-v` mount, or
  2. Container sets an `ENTRYPOINT` that performs configuration/initialisation, using some combination of environment variables, and or a default configuration file (most common for legacy applications), or
  3. The application can be configured in entirety with environment variables (container native app).

> no file mount for downloading

Either use host mounts or docker volumes. Or define one in the image an we just allow the user to chose.

> I'm a bit unsure of the value overall of providing these images, as they have to be a bit too opinionated (if you let the user specify the config for instance, the ports may change), and given the low dependency count of synapse probably don't provide much more value than compiling it yourself.

If you don't see value in providing an official docker image for end users, or even developers, then I'm not going to argue the point. Close this PR and I'll move on.

File: Dockerfile (hunk 1 above)
Comment (macropin) [reply Dockerfile:30]: Then that would a good approach for a completely optimised prod image. I'm not a rust developer. How does one determine the dependencies?

**Comment by Luminarys** [comment]
Closing this in favor of the other approach.

**Pull request closed without merging**